### ISSUE TITLE
resample_img segmentation fault

### ISSUE THREAD

**Issue by lesteve** [nilearn/nilearn#346]
Migrated from #322.

A simple example reproducing the problem:

```python

from nilearn.image import resample_img
import numpy as np
from nibabel import Nifti1Image


def test(fourth_dim):
    shape_in = (64, 64, 64)
    aff_in = np.diag([2., 2., 2., 1.])
    aff_out = np.diag([3., 3., 3., 1.])
    data = np.ones(shape_in + (fourth_dim, ), dtype=np.float64)

    img_in = Nifti1Image(data, aff_in)

    img_out = resample_img(img_in,
                           target_affine=aff_out,
                           interpolation='nearest')

    print 'shape: ', img_out.shape

test(1024)  # works fine on my machine
test(1025)  # segmentation fault
```

**Comment by banilo** [comment]
Cool!

**Comment by eickenberg** [comment]
I confirm -- 1024 works, 1025 segfaults. Thanks for isolating!

**Comment by eickenberg** [comment]
`np.log2(64 * 64 * 64 * 1024) == 28`

**Comment by lesteve** [comment]
I further tracked down the problem to scipy/ndimage/interpolation.py l.399 (scipy 0.14 on my machine)
```python

_nd_image.zoom_shift(filtered, matrix, offset, output, order,
                             mode, cval)
```

and a stand-alone example only using scipy:

```python

import numpy as np
from scipy import ndimage

def test(fourth_dim):
    shape_in =(64, 64, 64)
    shape_out =(43, 43, 43)
    arr_base = np.ones(shape_in + (fourth_dim,))
    arr_in = arr_base[..., 0]
    arr_out = np.zeros(shape_out)
    matrix = np.array([1.5, 1.5, 1.5])
    offset = np.array([0., 0., 0.])
    order = 0
    mode = 4
    cval = 0.

    ndimage._nd_image.zoom_shift(arr_in, matrix, offset, arr_out, order, mode, cval)
    print arr_out.mean()

test(1024)  # works fine
test(1025)  # segmentation fault

```

**Comment by eickenberg** [comment]
With `shape_in == (32, 64, 64)`, @lesteve and I both find segfault at 2115, works at 2114

**Comment by lesteve** [comment]
Here is the simplified snippet again without the power of 2 coincidence:

```python

import numpy as np
from scipy import ndimage

def test(fourth_dim):
    shape_in = (65, 65, 65)
    shape_out = (43, 43, 43)
    arr_base = np.ones(shape_in + (fourth_dim,))
    arr_in = arr_base[..., 0]
    arr_out = np.zeros(shape_out)
    matrix = np.array([1.5, 1.5, 1.5])
    offset = np.array([0., 0., 0.])
    order = 0
    mode = 4
    cval = 0.

    ndimage._nd_image.zoom_shift(arr_in, matrix, offset, arr_out, order, mode, cval)
    print arr_out.mean()

test(992)  # works fine
test(993)  # segmentation fault
```

**Comment by eickenberg** [comment]
With `shape_in == (64, 32, 64)` it segfaults at 2050 ...

**Comment by lesteve** [comment]
One thing of importance: all of the examples were run with scipy 0.14.0.

Some investigation with gdb showed indicated that the culprit was the int32 `idx` wrapping around and becoming negative causing a segmentation fault [there](https://github.com/scipy/scipy/blob/v0.14.0/scipy/ndimage/src/ni_interpolation.c#L869).

Looks like this has actually been fixed in scipy 0.14.1 through [this](https://github.com/scipy/scipy/pull/4218).  Although this doesn't seem related directly, the scipy PR changed the type of `idx` from int32 to int64, so that `idx` is a lot less likely to wrap around.

**Comment by dohmatob** [comment]
Nice reversing. Long live *gdb!*

On Mon, Jan 12, 2015 at 4:07 PM, Loïc Estève <notifications@github.com>
wrote:

> One thing of importance: all of the examples were run with scipy 0.14.0.
>
> Some investigation with gdb showed indicated that the culprit was the
> int32 idx wrapping around and becoming negative causing a segmentation
> fault there
> <https://github.com/scipy/scipy/blob/v0.14.0/scipy/ndimage/src/ni_interpolation.c#L869>
> .
>
> Looks like this has actually been fixed in scipy 0.14.1 through this
> <https://github.com/scipy/scipy/pull/4218>. Although this doesn't seem
> related directly, the scipy PR changed the type of idx from int32 to
> int64, so that idx is a lot less likely to wrap around.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/issues/346#issuecomment-69582909>.
>



-- 
DED
